internal/lock.go

```diff
@@ -49,30 +56,120 @@ func NewLock(path string, newOk bool) (*Lock, error) {
+func (l *Lock) AddResource(paths []string, algo string, tags []string, filename string, cacheURL string) error {
 	for _, u := range paths {
 		if l.Contains(u) {
 			return fmt.Errorf("resource '%s' is already present", u)
 		}
 	}
-	r, err := NewResourceFromUrl(paths, algo, tags, filename)
+	r, err := NewResourceFromUrl(paths, algo, tags, filename, cacheURL)
 	if err != nil {
 		return err
 	}
+
+	// If cache URL is provided, handles Artifactory upload
+	if cacheURL != "" {
+		token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+		if token == "" {
+			return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN environment variable is not set")
+		}
+
+		// Get file again for upload
+		path, err := GetUrltoTempFile(paths[0], context.Background())
+		if err != nil {
+			return fmt.Errorf("failed to get file for cache: %s", err)
+		}
+		defer os.Remove(path)
+
+		// Upload to Artifactory using hash as filename
+		err = uploadToArtifactory(path, cacheURL, r.Integrity)
+		if err != nil {
+			return fmt.Errorf("failed to upload to cache: %v", err)
+		}
+	}
+
 	l.conf.Resource = append(l.conf.Resource, *r)
 	return nil
 }
 
+func uploadToArtifactory(filePath, cacheURL, integrity string) error {
+	token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+	if token == "" {
+		return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN environment variable is not set")
+	}
+
+	// Read file
+	fileContent, err := os.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to read file: %v", err)
+	}
+
+	artifactoryURL := fmt.Sprintf("%s/%s", cacheURL, integrity)
+
+	// Create upload request
+	req, err := http.NewRequest("PUT", artifactoryURL, bytes.NewReader(fileContent))
+	if err != nil {
+		return fmt.Errorf("failed to create request: %v", err)
+	}
+
+	req.Header.Set("Authorization", "Bearer "+token)
+	req.Header.Set("Content-Type", "application/octet-stream")
+
+	// Do upload
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to upload: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("upload failed with status: %d", resp.StatusCode)
+	}
+
+	return nil
+}
+
 func (l *Lock) DeleteResource(path string) {
 	newStatements := []Resource{}
 	for _, r := range l.conf.Resource {
 		if !r.Contains(path) {
 			newStatements = append(newStatements, r)
+		} else if r.Contains(path) && r.CacheUri != "" {
+			token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+			if token == "" {
+				slog.Warn("Warning: Unable to delete from Artifcatory: GRABIT_ARTIFACTORY_TOKEN not set.")
+				continue
+			}
+
+			artifactoryURL := fmt.Sprintf("%s/%s", r.CacheUri, r.Integrity)
+
+			err := deleteCache(artifactoryURL, token)
+			if err != nil {
+				slog.Warn("Warning: Unable to delete from Artifcatory")
+			}
 		}
 	}
 	l.conf.Resource = newStatements
 }
 
+func deleteCache(url, token string) error {
+	req, err := http.NewRequest("DELETE", url, nil)
```

Comment:
Same remark here, let use `github.com/carlmjohnson/requests` for http interactions.